client/asset/btc/rpcclient.go

```diff
@@ -470,19 +473,153 @@ func (wc *rpcClient) signTx(inTx *wire.MsgTx) (*wire.MsgTx, error) {
 	return outTx, nil
 }
 
+func (wc *rpcClient) listDescriptors(private bool) (*listDescriptorsResult, error) {
+	descriptors := new(listDescriptorsResult)
+	return descriptors, wc.call(methodListDescriptors, anylist{private}, descriptors)
+}
+
 // privKeyForAddress retrieves the private key associated with the specified
 // address.
 func (wc *rpcClient) privKeyForAddress(addr string) (*btcec.PrivateKey, error) {
-	var keyHex string
-	err := wc.call(methodPrivKeyForAddress, anylist{addr}, &keyHex)
+	// Descriptor wallets do not have dumpprivkey.
+	if !wc.descriptors {
+		var keyHex string
+		err := wc.call(methodPrivKeyForAddress, anylist{addr}, &keyHex)
+		if err != nil {
+			return nil, err
+		}
+		wif, err := btcutil.DecodeWIF(keyHex)
+		if err != nil {
+			return nil, err
+		}
+		return wif.PrivKey, nil
+	}
+
+	// With descriptor wallets, we have to get the address' descriptor from
+	// getaddressinfo, parse out its key origin (fingerprint of the master
+	// private key followed by derivation path to the address) and the pubkey of
+	// the address itself. Then we get the private key using listdescriptors
+	// private=true, which returns a set of master private keys and derivation
+	// paths, one of which corresponds to the fingerprint and path from
+	// getaddressinfo. When the parent master private key is identified, we
+	// derive the private key for the address.
+	ai := new(GetAddressInfoResult)
+	if err := wc.call(methodGetAddressInfo, anylist{addr}, ai); err != nil {
+		return nil, fmt.Errorf("getaddressinfo RPC failure: %w", err)
+	}
+	wc.log.Tracef("Address %v descriptor: %v", addr, ai.Descriptor)
+	desc, err := dexbtc.ParseDescriptor(ai.Descriptor)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to parse descriptor %q: %w", ai.Descriptor, err)
+	}
+	// TODO: Do we need to support addresses for imported private keys that
+	// potentially have no key origin (master fingerprint and derivation path)?
```

Comment:
That's true only if `FundOrder` does not use one of the RPCs like getnewaddress.  AFAICT, the "active" flag marks them as the source to use for address generation.

If `getaddressinfo` returned a descriptor with no key origin, just a bare private key, the way to handle that would be to just search through the `listdescriptors` result for a matching bare private key, so I think it's straight forward.  I tried to test this and that's when I found out about the "active" rule.

`./desc importdescriptors '[{"desc":"wpkh(cSqGdZNwiMcqJqC1NCwigLtQWmooMNnz5jMVuzuLd4pRiLP7CgFM)#r3wwy3fg", "timestamp": 1655220035, "active": false}]'`

You can only import a bare WIF private key like above if "active" is false.  I set it to true and it denied the request saying only ranged descriptors can be active (needed for HD key generation at a range child indexes).  So that's when I concluded it was pointless to support if we cannot request such addresses.

I can implement the bare key matching anyway, but it I didn't see it ever hitting.  I might be reasoning this out incorrectly.

Reply:
I imported a base WIF private key descriptor and send funds to it. These were the only funds in the wallet. Then when I tried to make a trade, it failed due to this error in `SignMessage`. Of course I don't think anyone would actually do this, but it's possible.

Reply:
Ohhh it's because we use listunspent to fund the order not getnewaddress.
Good catch.